### DISCUSSION TITLE
How to add different BMS to the drive?

### DISCUSSION THREAD

**Discussion by stanhausc** [Louisvdw/dbus-serialbattery#97] [Q&A] [answered]
Hello, I'm a mechanical engineer and due to the lack of professionals in the area I'm working as a scholarships in the development of drivers.

I work at [PTI](https://www.pti.org.br/), here we have numerous renewable energy projects, and one of them is the development of communication between this [BMS](https://wdcnet.com.br/wp-content/uploads/2021/10/Protocolo-de-comunicacao-LFP-MODBUS_-rev.05.pdf) and Venus OS.

I'm studying your drive, and as I'm not from the area I apologize for the simple questions.

1- Could you show me the way for me to add another [BSM ](https://wdcnet.com.br/wp-content/uploads/2021/10/Protocolo-de-comunicacao-LFP-MODBUS_-rev.05.pdf) to your drive?

   1.1 - Start with: battery_template.py (editing SOC, voltage, current, etc)?

           and also dbus-serialbattery.py on the lines:

          16 from dbushelper import DbusHelper
          17 from utils import DRIVER_VERSION, DRIVER_SUBVERSION, logger

  1.2 - The BSM document is a bit simple, I believe it still lacks information to develop the driver.

2- I have another drive developed for FIAMM batteries, but I found it more confusing than yours, if you want I can share.

3- I'm more interested in learning how to develop than having the drive ready!

Thanks for sharing your great work.

**Comment by Louisvdw** [comment, marked as answer]
Hi @stanhausc 
You should start by making a copy of the battery_template.py and naming it for your BMS. The template is a small example implementation. In theory this is the only file you will need to edit to add your BMS, except when adding this new file as a reference to the main program. 

Rename the BatteryTemplate class for your BMS and then you need to use the 3 main functions that are implemented to read and populate data for your BMS. (test_connection, get_settings & refresh_data)
These functions are called at different times.

There is a helper function at the bottem that you can use to write the commands and read your data. There are small differences in each comms protocol and this is where the parameters to read_serial_data needs to be adjusted for your BMS. 

To start you only need the number of cells, the soc, current and voltage. Once you can read this the driver should already be working for you and you can add the rest of the items. If you populate all the variables in the battery class via your derived BMS class then the rest of the driver will use that to push the data to VenusOS. 

I do suggest you make a fork of the repo and then create a new branch inside that fork. If you are ready you can then create a pull request of your changes to be included back into the main repo and once it is all ready and tested by a few volenteers we will merge the branch back into the main branch and release.

**Comment by Louisvdw** [comment]
PS. Have a look at the other battery bms implementations. Many of them share similar coding pieces and you will see examples of how they are implemented. They are most MODBUS RTU based same as your BMS. 
Start with just a master and the command will be structured as in 1.3.1. The address is the list in 3.
LENGTH_POS will be the byte position of your Byte Cound field and LENGTH_CHECK is how many bytes does all the stuff outside the data packet are. 

You can find a list of the other BMS around lines 42-50 in dbus-serialbattery.py

**Reply by stanhausc** [reply]
Hello @Louisvdw, Sorry for asking more questions.

 I'm having difficulties with the CRC.

At first I made changes as you guided me by creating my BatteryTemplate:

`class LifePo4(Battery):

    def __init__(self, port,baud):
        super(LifePo4, self).__init__(port,baud)
        self.type = self.BATTERYTYPE
        
    logger.warning('Testing Hello word')
    BATTERYTYPE = "LifePo4"
    LENGTH_CHECK = 4
    LENGTH_POS = 3
    command_general = b"\x01\x03\x00\x00\x00\x7C\x44\x03"`

![image](https://user-images.githubusercontent.com/85182432/159949627-353107a9-cd90-4d79-b749-b74f939c0935.png)

command_general =
x01 address slave1
x03 to reading logs
x00 x00 reading start address
x00 x7C to read all registers (124 in total)

![image](https://user-images.githubusercontent.com/85182432/159952791-f37ad0b0-be55-459f-a4d9-8f2d9fc19bb3.png) 
![image](https://user-images.githubusercontent.com/85182432/159952863-04a35c1a-ef11-4bae-9628-359355e35646.png)

x44 x03 is the calculation of the crc, inverted.

I added my template to the dbus-serialbattery.py file.

I'm having trouble understanding this part of the code:

`    def read_serial_data_jkbms(self, command):
        # use the read_serial_data() function to read the data and then do BMS spesific checks (crc, start bytes, etc)
        data = read_serial_data(command, self.port, self.baud_rate, self.LENGTH_POS, self.LENGTH_CHECK,None, self.LENGTH_SIZE)
        if data is False:
            return False

        start, length = unpack_from('>HH', data)
        terminal = unpack_from('>L', data[4:])[0]
        cmd, src, tt = unpack_from('>bbb', data[8:])
        frame = unpack_from('>H', data[-9:])[0]
        end, crc_hi, crc_lo = unpack_from('>BHH', data[-5:])

        s = sum(data[0:-4])
        logger.debug('S%d C%d C%d' % (s, crc_lo, crc_hi))
        logger.debug('T%d C%d S%d F%d TT%d' % (terminal, cmd, src, frame, tt))

        if start == 0x4E57 and end == 0x68 and s == crc_lo:
            return data[10:length-19]
        elif s != crc_lo:
            logger.error('CRC checksum mismatch: Expected 0x%04x, Got 0x%04x' % (crc_lo, s))
            return False
        else:
            logger.error(">>> ERROR: Incorrect Reply ")
            return False`

Most templates use Checksum instead of CRC.

What is the meaning of the letters 'BBB' , 'H' , '>H', '>BHH' ? 
B:unsigned byte?
H:unsigned short?

why starts at 0x4e57 and ends at 0x68?

If you can explain this part of the code to me when you have some time.

Thank's.
